pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,223 @@
+package delegatedidentity
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	delegatedidentityv1 "github.com/spiffe/spire-api-sdk/proto/spire/api/agent/delegatedidentity/v1"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+
+	attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/endpoints"
+	"github.com/spiffe/spire/pkg/agent/manager"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/pkg/common/x509util"
+	"github.com/spiffe/spire/pkg/server/api"
+	"github.com/spiffe/spire/proto/spire/common"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RegisterService registers privileged service on provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	delegatedidentityv1.RegisterDelegatedIdentityServer(s, service)
+}
+
+type Attestor interface {
+	Attest(ctx context.Context) ([]*common.Selector, error)
+}
+
+type Config struct {
+	Log             logrus.FieldLogger
+	Manager         manager.Manager
+	Attestor        attestor.Attestor
+	AuthorizedUsers []string
+}
+
+func New(config Config) *Service {
+	authorizedUsers := map[string]bool{}
+
+	for _, user := range config.AuthorizedUsers {
+		authorizedUsers[user] = true
+	}
+
+	return &Service{
+		manager:         config.Manager,
+		attestor:        endpoints.PeerTrackerAttestor{Attestor: config.Attestor},
+		authorizedUsers: authorizedUsers,
+	}
+}
+
+// Service implements the privileged server
+type Service struct {
+	delegatedidentityv1.UnsafeDelegatedIdentityServer
+
+	manager  manager.Manager
+	attestor Attestor
+
+	// Spiffe IDs of users that are authorized to use this API
+	authorizedUsers map[string]bool
+}
+
+// isCallerAuthorized attests the caller and returns true if its identity is on
+// the authorized users map.
+func (s *Service) isCallerAuthorized(ctx context.Context) (bool, error) {
+	callerSelectors, err := s.attestor.Attest(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	identities := s.manager.MatchingIdentities(callerSelectors)
+
+	for _, identity := range identities {
+		id := identity.Entry.SpiffeId
+
+		if _, ok := s.authorizedUsers[id]; ok {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (s *Service) SubscribeToX509SVIDs(req *delegatedidentityv1.SubscribeToX509SVIDsRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509SVIDsServer) error {
+	ctx := stream.Context()
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		return fmt.Errorf("failed to attest caller: %w", err)
```

Comment:
Since this is grpc handler code, we'll want to return grpc status code on all of the returns in this method.

Additionally, consideration should be made with regards to information leakage. I don't know what the potential failure modes are here, but it may be better to log the error details and return something less detailed to the caller

e.g. log `failed to attest caller: %w` and return `status.Error(codes.Internal, "workload attestation failure")`

All errors/returns in these public handler methods should be looked at through this lens.

Reply:
Good,

I just saw how it's done in workloadAPI [FetchX509SVID](https://github.com/spiffe/spire/blob/main/pkg/agent/endpoints/workload/handler.go#L221)
`s.isCallerAuthorized(ctx)` just returns the error from `s.attestor.Attest(ctx)` that can be `"peer tracker watcher missing from context"` or `"could not verify existence of the original caller: %v"`. So following your suggestion, should be like the follows, right?
```
if err != nil {
    log.WithError(err).Error("Workload attestation failed")
    return err
}
```

Reply:
> I just saw how it's done in workloadAPI FetchX509SVID
s.isCallerAuthorized(ctx) just returns the error from s.attestor.Attest(ctx) that can be "peer tracker watcher missing from context" or "could not verify existence of the original caller: %v". 

aay ... sorry about that. No it is not the best thing ... I think better would be:

```
if err != nil {
    log.WithError(err).Error("Workload attestation failed")
    return status.Error(codes.Internal, "workload attestation failed")
}
```

it's ok to return more detailed error data to the caller in cases where we understand exactly the kind of stuff that might be returned ... but code changes so IMO the safe thing to do is to be terse about it.

If you return an error without using grpc status code, it defaults to codes.Internal

Reply:
I was just editing when you send the response ahhaha I think we think the same way

Reply:
FWIW, `codes.Unknown` is the default if a non-status error is returned from a handler. I think the choice of `codes.Internal` as @evan2645 suggested is appropriate. We should definitely be returning explicit status errors from the handlers (there is older code that doesn't do this, but all new code should).

Reply:
Folks, following some other points, should we be less detail when return to user, when the caller is not authorized ?
From:
```
if !authorized {
    return status.Error(codes.PermissionDenied, "no authorized")
}
```
To:
```
if !authorized {
    log.Error(codes.PermissionDenied, "no authorized")
    return status.Error(codes.Internal, "workload attestation failed")
}
```

Reply:
Folks, for `FetchX509SVID` can be returned `status.Error(codes.PermissionDenied, "no identity issued")` at some point. Because in our case the API is privileged, would be better to return just `status.Error(codes.PermissionDenied, "no identity issued")` or just follow my last example `status.Error(codes.Internal, "workload attestation failed")`?

https://github.com/spiffe/spire/blob/1c2470a80a04c9c8eb15077cb12cb7c70adbf262/pkg/agent/endpoints/workload/handler.go#L308-L314